### PR TITLE
Add Finite State Machine for cleaner flow of logic

### LINKED ISSUE #4: Not detecting OpenIn State
#2 - Issue: During door testing states openin (-1) has not been detected. 
- the state has not been detected during a flap
- the state has not been detected holding the magnet manually
- the state does not print, and the LED does not show, it continued to show closed state with a blue light
# Sample data from testing
```0
1
0
1
0
1
0
1
0
1
0
1
0
1
0
1
0
1
0
1
0
1

### PR DESCRIPTION
**Pull request by tjquillan**
This PR transitions state management to a Finite State Machine that can hooked into by external functions. This makes the management of much cleaner in the main loop.

### CODE REVIEW

**Review by atramirez — APPROVED** [review]
Nothing I found wrong with the logic, we will need to test extensively on next meeting.

File: lib/fsm/door_fsm.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#ifndef PET_PASSAGE_FSM_DOOR_FSM_H_
```
Comment (atramirez) [inline lib/fsm/door_fsm.hpp:1]: For your #ifndef and #define there is an inconsistency. In the files under lib/hardware you use 
 `#ifndef PET_PASSAGE_LIB_MAGNETIC_SENSOR_H_ `
Noting LIB here but not hardware subfile, though in these files you do not cite LIB but FSM. Review the style guide for the right form for consistency?

File: lib/fsm/door_fsm.ipp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#ifndef PET_PASSAGE_FSM_DOOR_FSM_I_
```
Comment (atramirez) [inline lib/fsm/door_fsm.ipp:1]: Same inconsistency as the previous file.

**Review by atramirez — COMMENTED** [review]
Found a small inconsistency, nothing groundbreaking, but let's check the style guide, and check our files when we add folders.

File: lib/fsm/door_fsm.hpp (hunk 1 above)
Comment (tjquillan) [reply lib/fsm/door_fsm.hpp:1]: Based on the example it provides of `foo/src/bar/baz.h` becoming `FOO_BAR_BAZ_H_` I believe the presence of LIB was a mistake in `PET_PASSAGE_LIB_MAGNETIC_SENSOR_H_`. However, this should be addressed in a seperate PR.

File: lib/fsm/door_fsm.ipp (hunk 2 above)
Comment (tjquillan) [reply lib/fsm/door_fsm.ipp:1]: see: https://github.com/Pet-Passage/Pet-Passage/pull/2#discussion_r671465045

**Pull request merged**